### DISCUSSION TITLE
Go modules: track HEAD of default branch rather than latest tag?

### DISCUSSION THREAD

**Discussion by dnwe** [renovatebot/renovate#15927] [Request Help]
For certain Go dependencies, it may be desirable to track the HEAD commit of the default branch rather than tracking against released (tagged) versions. Has anyone done this and can they share their renovate.json snippet to do so?

e.g., on a certain application I might like renovate to perform the equivalent of `go get github.com/Shopify/sarama@HEAD` rather than `go get github.com/Shopify/sarama@latest` and hence bump to the pseudo version based off the latest commit (v1.34.1-0.20220607101826-9bf344fc6f90 at time of writing) rather than the "latest" tagged release of v1.34

**Comment by rarkins** [comment]
I'm not a Go user, so excuse me if I misunderstand anything.

Getting the HEAD commit might mean that there's *no* tag associated. In such case does it revert to being a `0.0.0-` pseudo version or is it Go's behavior that it will somehow reuse whatever was the last tag in the git history? What if that tag isn't a semver?

If the latest commit is tagged then why is there a need to use a pseudo version instead of simply the tag?

**Reply by dnwe** [reply]
@rarkins Go’s behaviour is as shown in the example above, it if there are newer commits than the latest published tag then it will bump the patch part of semver and append a hyphen then the commit date and sha1. 

The desire here is to specifically track unreleased versions. I believe that “go mod tidy” will convert a plain sha1 to this pseudo versioning too, so renovate could use git versioning and the git repo as the source to update go.mod, and then the go mod tidy will sort things.

**Reply by rarkins** [reply]
Sounds like you want the equivalent of Renovate's "ignoreUnstable" feature?

**Reply by dnwe** [reply]
@rarkins not exactly — as I understand it [ignoreUnstable](https://docs.renovatebot.com/configuration-options/#ignoreunstable) would permit upgrading to non-semver release tags assuming they were tagged in the git repo and you were using [releases-direct](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/go/releases-direct.ts) or they conformed to the [pre-release format ](https://go.dev/doc/modules/release-workflow#pre-release) such that the [releases-goproxy.ts](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/go/releases-goproxy.ts) method would find them.

So I think currently this is not possible with renovate because [releases-direct](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/go/releases-direct.ts) only supports fetching "releases" in github and [releases-goproxy.ts](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/go/releases-goproxy.ts) only supports making a /@v/list query for versions to the GOPROXY

Whilst you could attempt to knock something up with a regexManager to put a git ref into go.mod along the lines of:
```
# renovate: datasource=git-refs depName=https://github.com/Shopify/sarama versioning=git
```

I don't think you would get the gomodTidy from postUpdateOptions so you wouldn't get the correct pseudo version from the go toolchain

**Reply by rarkins** [reply]
Let's say Renovate achieved what you want and then you are then using `v1.34.1-0.20220607101826-9bf344fc6f90`. Now let's say the source directory updates to `v1.34.2`. If Renovate just inserted the plain sha1 corresponding to `v1.34.2` then would `go mod tidy` resolve that to `v1.34.2`? i.e. could Renovate just focus on the sha's and let `go` take care of whether they matcha tag or not? I'm not sure that would be ideal though because Renovate would say "Update Shopify/sarama digest to 263a9fcc12" in its title if it's unaware of the tags
